src/App.tsx

```diff
@@ -1,26 +1,213 @@
-/* eslint-disable max-len */
-/* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
-import { UserWarning } from './UserWarning';
-
-const USER_ID = 0;
+import React, { useEffect, useMemo, useState } from 'react';
+import { Header, Footer, Notification, TodoList } from './components';
+import {
+  USER_ID,
+  getTodos,
+  addTodo,
+  deleteTodo,
+  updateTodo,
+} from './api/todos';
+import { FilterOption, Todo, Error } from './types';
+import { useError } from './hooks';
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filterOption, setFilterOption] = useState<FilterOption>(
+    FilterOption.all,
+  );
+  const [todoInputValue, setTodoInputValue] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isLoading, setIsLoading] = useState(false);
+  const [loadingTodoIds, setLoadingTodoIds] = useState<number[]>([]);
+  const [editingTodoId, setEditingTodoId] = useState<number | null>(null);
+  const [focusHeader, setFocusHeader] = useState(true);
+  const { error, handleError, handleResetError } = useError();
+
+  const todosAmount = useMemo(() => todos.length, [todos]);
+  const uncompletedTodosAmount = useMemo(
+    () => todos.reduce((amount, todo) => amount + (todo.completed ? 0 : 1), 0),
+    [todos],
+  );
+
+  const handleSubmitTodo = (title: string) => {
+    if (!title) {
+      handleError(Error.EMPTY_TITLE);
+
+      return;
+    }
+
+    const newTodo = {
+      title: title.trim(),
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo({ ...newTodo, id: 0 });
+
+    setIsLoading(true);
+    setFocusHeader(false);
+
+    addTodo(newTodo)
+      .then(response => {
+        setTodos(currentTodos => [...currentTodos, response]);
+        setTodoInputValue('');
+      })
+      .catch(() => {
+        handleError(Error.ADDING_TODO);
+      })
+      .finally(() => {
+        setIsLoading(false);
+        setTempTodo(null);
+        setFocusHeader(true);
+      });
+  };
+
+  const handleDeleteTodo = (id: number) => {
+    setIsLoading(true);
+    setLoadingTodoIds(current => [...current, id]);
+    setFocusHeader(false);
+
+    deleteTodo(id)
+      .then(() => {
+        setTodos(currentTodos => currentTodos.filter(todo => todo.id !== id));
+
+        if (editingTodoId) {
+          setEditingTodoId(null);
+        }
+      })
+      .catch(() => handleError(Error.DELETING_TODO))
+      .finally(() => {
+        setLoadingTodoIds(current =>
+          current.filter(deletedId => deletedId !== id),
+        );
+        setIsLoading(false);
+        if (!editingTodoId) {
+          setFocusHeader(true);
+        }
+      });
+  };
+
+  const handleDeleteCompleted = () => {
+    const completedIds = todos.reduce<number[]>((ids, todo) => {
+      if (todo.completed) {
+        ids.push(todo.id);
+      }
+
+      return ids;
+    }, []);
```

Comment:
Use `filter `to create an array of completed todos, then `map `to extract their IDs. This improves readability and keeps the code straightforward.
```suggestion
const completedIds = todos
    .filter(todo => todo.completed)
    .map(todo => todo.id);
```